hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -1225,130 +1301,292 @@ private boolean innerRename(Path source, Path dest)
-          removeKeys(keysToDelete, true, false);
-        }
+
+        // await the final set of copies and their deletion
+        // This will notify the renameTracker that these objects
+        // have been deleted.
+        completeActiveCopiesAndDeleteSources.apply("final copy and delete");
+
+        // We moved all the children, now move the top-level dir
+        // Empty directory should have been added as the object summary
+        renameTracker.moveSourceDirectory();
       }
-      if (!keysToDelete.isEmpty()) {
-        removeKeys(keysToDelete, false, false);
+    } catch (AmazonClientException | IOException ex) {
+      // rename failed.
+      // block for all ongoing copies to complete, successfully or not
+      try {
+        completeActiveCopies.apply("failure handling");
+      } catch (IOException e) {
+        // a failure to update the metastore after a rename failure is what
+        // we'd see on a network problem, expired credentials and other
+        // unrecoverable errors.
+        // Downgrading to warn because an exception is already
+        // about to be thrown.
+        LOG.warn("While completing all active copies", e);
       }
 
-      // We moved all the children, now move the top-level dir
-      // Empty directory should have been added as the object summary
-      if (hasMetadataStore()
-          && srcPaths != null
-          && !srcPaths.contains(src)) {
-        LOG.debug("To move the non-empty top-level dir src={} and dst={}",
-            src, dst);
-        S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, src, dst,
-            username);
-      }
+      // notify the rename tracker of the failure
+      throw renameTracker.renameFailed(ex);
     }
 
-    metadataStore.move(srcPaths, dstMetas);
+    // At this point the rename has completed in the S3 store.
+    // Tell the metastore this fact and let it complete its changes
+    renameTracker.completeRename();
 
     if (!src.getParent().equals(dst.getParent())) {
       LOG.debug("source & dest parents are different; fix up dir markers");
       deleteUnnecessaryFakeDirectories(dst.getParent());
       maybeCreateFakeParentDirectory(src);
     }
-    return true;
+    return bytesCopied.get();
+  }
+
+  /**
+   * Remove source objects and update the metastore by way of
+   * the rename tracker.
+   * @param renameTracker rename tracker to update.
+   * @param keysToDelete list of keys to delete
+   * @param pathsToDelete list of paths matching the keys to delete 1:1.
+   * @throws IOException failure
+   */
+  @Retries.RetryMixed
+  private void removeSourceObjects(
+      final RenameTracker renameTracker,
+      final List<DeleteObjectsRequest.KeyVersion> keysToDelete,
+      final List<Path> pathsToDelete)
+      throws IOException {
+    List<Path> undeletedObjects = new ArrayList<>();
+    try {
+      // remove the keys
+      // this does will update the metastore on a failure, but on
```

Comment:
nit: /does will/does/

Reply:
another nit: the asymmetry here in metadatastore updating (only on failure) feels a bit odd

Reply:
I know, it grates on me too. But the renameTracker is invoked on success, so it is all contained in this operation.